File: binary search/first_bad_version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        l=1
        r=n
        fbv=0
        
        while l<=r :
        
            mid=(l+r)//2;
            x = isBadVersion(mid);
            
            if x==True:
            
                r=mid-1
                fbv = mid
            
            
            else:
            
                l=mid+1
        
        return fbv

File: binary search/first_bad_version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int l=1, r=n;
        int fbv=0;
        
        while(l<=r)
        {
            int mid=l+(r-l)/2;
            boolean x = isBadVersion(mid);
            
            if(x==true)
            {
                r=mid-1;
                fbv = mid;
            }
            
            else
            {
                l=mid+1;
            }
        }
        
        return fbv;
    }
}